torch/csrc/jit/autodiff.cpp

```diff
@@ -34,38 +45,233 @@ static std::vector<Value*> gradientForNode(Node* node, ArrayRef<Value*> grad_val
-  JIT_ASSERT(graph->stage() == 0);
-  graph->advanceStage();
+// Before:
+//   - graph has only stage 0
+//   - grad_desc doesn't have any fields initialized
+// After:
+//   - graph has stage 0 and stage 1 that computes its vjp
+//   - grad_desc has df_input_vjps and df_output_vjps set
+//     (but df_input_vjps will be modified later as well)
+static value_map addReverseInline(Graph& graph, Gradient& grad_desc) {
+  JIT_ASSERT(graph.stage() == 0);
+  graph.advanceStage();
+
+  value_map grad_map; // x -> dx mapping
+  const auto get_grad = [&](Value* v) { return grad_map.at(v); };
+  const auto set_grad = [&](Value *x, Value *dx) {
+    if (Value * prev_grad = grad_map[x]) {
+      Value * new_grad = addAndPutAfter(prev_grad, dx, dx->node());
+      grad_map[x] = new_grad;
+    } else {
+      grad_map[x] = dx;
+    }
+  };
 
-  std::unordered_map<Value*, Value*> grad_map; // x -> dx mapping
-  const auto get_grad = [&](Value* v) { return grad_map[v]; };
-  for (auto output : graph->outputs())
-    grad_map[output] = graph->addInput()->setType(output->typeOption());
+  auto outputs = graph.outputs();
+  for (std::size_t i = 0, num_outputs = outputs.size(); i < num_outputs; ++i) {
+    Value * output = outputs[i];
+    Value * output_grad = graph.addInput()->setType(output->typeOption());
+    set_grad(output, output_grad);
+    grad_desc.df_input_vjps.push_back(i);
+  }
 
-  for (auto it = graph->rbegin(), end = graph->rend(); it != end; ++it) {
+  for (auto it = graph.rbegin(), end = graph.rend(); it != end; ++it) {
     Node *node = *it;
     auto inputs = node->inputs();
     value_list grad_inputs = gradientForNode(node, fmap(node->outputs(), get_grad));
     JIT_ASSERT(grad_inputs.size() == node->inputs().size());
     for (std::size_t i = 0, num_inputs = grad_inputs.size(); i < num_inputs; ++i) {
-      if (Value * prev_grad = grad_map[inputs[i]]) {
-        Node *new_grad_node = graph->create(kadd, {prev_grad, grad_inputs[i]})
-                                   ->t_(kalpha, at::Scalar(1).toTensor());
-        new_grad_node->insertAfter(grad_inputs[i]->node());
-        Value *new_grad = new_grad_node->output();
-        new_grad->setType(prev_grad->typeOption());
-        grad_map[inputs[i]] = new_grad;
-      } else {
-        grad_map[inputs[i]] = grad_inputs[i];
-      }
+      set_grad(inputs[i], grad_inputs[i]);
     }
   }
 
-  for (auto input : graph->inputs()) {
+  auto inputs = graph.inputs();
+  for (std::size_t i = 0, num_inputs = inputs.size(); i < num_inputs; ++i) {
+    Value * input = inputs[i];
     if (input->stage() > 0) break;
-    graph->registerOutput(grad_map.at(input));
+    graph.registerOutput(get_grad(input));
+    grad_desc.df_output_vjps.push_back(i);
+  }
+
+  return grad_map;
+}
+
+// This function will take the graph and return a new one that:
+//   - contains all nodes of graph that have given stage
+//   - there will be an input corresponding to each input of the inputs array
```

Comment:
I suppose it would clarify if you mentioned that an 'input' doesn't necessarily have to be an input of the graph itself; it may be a temporary from a previous stage you want to flow in.